NumLib/TimeStepping/Algorithms/EvolutionaryPIDcontroller.cpp

```diff
@@ -104,6 +109,28 @@ bool EvolutionaryPIDcontroller::next(const double solution_error)
     return true;
 }
 
+double EvolutionaryPIDcontroller::limitStepSize(
+    const double h_new, const double h_n,
+    const bool previous_step_accepted) const
+{
+    const double h_in_range = std::max(_h_min, std::min(h_new, _h_max));
+    double limited_h =
+        std::max(_rel_h_min * h_n, std::min(h_in_range, _rel_h_max * h_n));
+
+    // If the last time step was rejected and the new time step size is
+    // identical to that of the previous rejected step, the new step size is
+    // then reduced by half.
+    if (!previous_step_accepted)
+    {
+        if (std::fabs(limited_h - _ts_current.dt()) <
+            std::numeric_limits<double>::min())
+            limited_h = std::max(_h_min, 0.5 * limited_h);
+        if (limited_h > _ts_current.dt())
+            limited_h = std::max(_h_min, 0.5 * _ts_current.dt());
```

Comment:
Not sure if the relative growth/shrinkage rates are checked. Could you please elaborate a little bit more?